Python Help Window Prototype to remove qtassistant

darshdinger opened the pull request:
### Description of work

#### Summary of work
<!-- Please provide a short, high level description of the work that was done.
-->
These enhancements feature refined CMake and Conda configurations alongside innovative implementations in both C++ and Python for a modern, Python-based help window. Collectively, these improvements lay a solid foundation for migrating away from the legacy QtAssistant system toward a more contemporary, flexible approach that leverages web based tech. This transition is designed to resolve longstanding challenges, such as the excessive package size and the absence of JavaScript support.
<!-- Why has this work been done? If there is no linked issue please provide appropriate context for this work.
#### Purpose of work
This can be removed if a github issue is referenced below
-->
The primary purpose of this work was to enhance the user documentation experience in Mantid Workbench by:

* Reducing dependency on the bulky QtAssistant system.
* Preparing the system for future integration with QtWebEngine/QtWebView to support modern web features like JavaScript for dynamic content and equation rendering.
* Bridging Python and C++ layers for a seamless help interface that can dynamically switch between local and online documentation sources.

Ref #37248  <!-- and fix #xxxx or close #xxxx xor resolves #xxxx. One line per issue fixed. -->
<!-- alternative
*There is no associated issue.*
-->

<!-- If the original issue was raised by a user they should be named here. Do not leak email addresses
**Report to:** [user name]
-->

#### Further detail of work
<!-- Please provide a more detailed description of the work that has been undertaken.
-->
CMake and Build Configuration:

* Updated the top-level `CMakeLists.txt` and associated Conda recipe files to accommodate new build options and dependencies (thanks @peterfpeterson). These changes ensure that necessary packages like `PyQtWebEngine` are available in the environment, allowing for `QtWebView` integration.

Help Window Architecture Enhancements:
* **Python Integration**:
   - Implemented `PythonHelpBridge` to establish a communication channel between C++ and Python. This bridge initializes the Python interpreter, imports the necessary Python modules, and calls Python functions to display help pages.
   - Developed `PythonHelpWindow` class which utilizes the bridge to invoke Python functions for displaying various documentation pages (algorithms, concepts, fit functions, and custom interfaces). This paves the way for using QtWebEngine in future iterations by abstracting the method of displaying web content.
   - Integrated logic in the help window presenter and view to decide between loading local documentation and falling back to the online Mantid documentation based on environment variables (`MANTID_LOCAL_DOCS_BASE`) and file existence.
   - Employed `QWebEngineView` (as seen in the Python samples and bridging code) to provide a more modern method of rendering documentation. Although the full QtWebEngine integration is prototyped and contextually referenced, the current changes establish the foundational elements necessary for this transition.

* **C++ Components**:
   - Refactored the `MantidHelpWindow` to improve how the help system locates and loads the QtHelp collection file (`MantidProject.qhc`).
   - Enhanced the `InterfaceManager`'s functionality to dynamically manage help window instances. Depending on build configurations (e.g., using `cmake .. -DDOCS_QTHELP=OFF`), it can intelligently instantiate either a native C++ help window or a Python-based alternative.

### To test:

This is suggested things to test. Trying to start the help from other methods is suggested as well. 
1. Rather than creating a new mantid-developer conda package, install `pyqtwebengine` on top of your existing mantid-developer conda environment. You may want to pin the version of pyqtwebengine to whatever version of pyqt you're using so conda doesn't uninstall mantid-developer
2. Re-run cmake with a switch to use the qtwebengine implementation `cmake . -DDOCS_QTHELP=off`
3. Build mantidworkbench and start it
4. Press `F1` or use the "Help"->"Mantid Help" menu item. This should bring up the new help window.
5. Open any algorithm and press the "?" button to bring up the help.
6. Use the menu to open "Interfaces"->"Diffraction"->"Engineering Diffraction" then press the "?" button to bring up the help.

Also try both leaving the help open and starting a different help page and closing the help and starting a different help page.

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
If you add release notes please save them as a separate file using the Issue or PR number as the file name. Check the file is located in the correct directory for your note(s).
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

darshdinger committed:
Changes to facilitate new HelpWindow design

darshdinger committed:
Changes to facilitate new HelpWindow design

Some updates to make branch current, tests in next

Most recent changes.

Updating branch with latest changes

Make qtassistant a compile time option

This removes the pointer to the qtassistant wrapper as dependent on
the variable QT_DOCS. The default behavior is to keep the current
functionality.

Hopefully this is a compile time implementation

Smal fixes

Final updates

More updates and Pete's changes

Revert some of Petes changes and some more updates

Reverted changes.

Some more changes to hopefully fix build issue

Another update to PythonHelpWindow

Get the code to compile

Updates to InterfaceManager (remove extra stuff)

Messed up in my last changes to InterfaceManager

Change how module is written in code

Fix pre-processor macro

Add pyqtwebengine dependency for new help window

Some fixes, seg fault investigation on local

Updates to fix cppcheck.

...

...

darshdinger committed:
Some updates to make branch current, tests in next

darshdinger committed:
Most recent changes.

darshdinger committed:
Updating branch with latest changes

peterfpeterson commented on qt/widgets/mplcpp/inc/MantidQtWidgets/MplCpp/Colormap.h:
```diff
@@ -11,7 +11,11 @@
 #undef slots
 #include "MantidQtWidgets/Common/Python/Object.h"
 #pragma pop_macro("slots")
+#include <QPushButton>
```
Why did you add all of these imports in this file? `Colormap.cpp` didn't change in this PR so I'm surprised that more imports are needd.

peterfpeterson commented on qt/widgets/common/CMakeLists.txt:
```diff
@@ -1,3 +1,7 @@
+# Add the Python scripts directory to the build
+set(PYTHON_SCRIPTS_DIR ${CMAKE_SOURCE_DIR}/mantid/qt/python/mantidqt/mantidqt/widgets/helpwindow)
```
This should be done in the pyproject.toml or somewhere similar. I don't think that it will work once packaged.

peterfpeterson commented on qt/widgets/common/inc/MantidQtWidgets/Common/HelpWindow.h:
```diff
@@ -8,28 +8,59 @@
+ * help pages, algorithm details, and concepts. It supports both
+ * the C++ HelpWindow and launching a Python-based HelpWindow.
+ */
+class EXPORT_OPT_MANTIDQT_COMMON HelpWindow : public QWidget {
```
What effect does dropping the inheritence from `QWidget` have on available methods/macros?

peterfpeterson committed:
Make qtassistant a compile time option

This removes the pointer to the qtassistant wrapper as dependent on
the variable QT_DOCS. The default behavior is to keep the current
functionality.

darshdinger committed:
Hopefully this is a compile time implementation

darshdinger committed:
Smal fixes

darshdinger committed:
Final updates

darshdinger committed:
More updates and Pete's changes

peterfpeterson commented on qt/widgets/common/src/PythonHelpWindow.cpp:
```diff
@@ -0,0 +1,90 @@
+namespace MantidQt {
+namespace MantidWidgets {
+
+#ifndef DOCS_QTHELP
```
You missed turning off `REGISTER_HELPWINDOW(HelpWindow)` in the `HelpWindow.cpp` based on the other case of this macro

darshdinger committed:
Revert some of Petes changes and some more updates

darshdinger committed:
Reverted changes.

darshdinger committed:
Some more changes to hopefully fix build issue

darshdinger committed:
Another update to PythonHelpWindow

peterfpeterson committed:
Get the code to compile

darshdinger committed:
Updates to InterfaceManager (remove extra stuff)

darshdinger committed:
Messed up in my last changes to InterfaceManager

peterfpeterson committed:
Change how module is written in code

peterfpeterson committed:
Fix pre-processor macro

peterfpeterson committed:
Add pyqtwebengine dependency for new help window

darshdinger committed:
Some fixes, seg fault investigation on local

darshdinger committed:
Updates to fix cppcheck.

darshdinger committed:
...

darshdinger committed:
...

darshdinger committed:
fixup! ...

darshdinger committed:
fixup! Merge branch '37248-remove-qtassistant-and-reduce-size-of-mantiddocs-package' of https://github.com/mantidproject/mantid into 37248-remove-qtassistant-and-reduce-size-of-mantiddocs-package

darshdinger committed:
Merge branch '37248-remove-qtassistant-and-reduce-size-of-mantiddocs-package' of https://github.com/mantidproject/mantid into 37248-remove-qtassistant-and-reduce-size-of-mantiddocs-package

darshdinger committed:
Updates to CMakeLists.txt for cppcheck

darshdinger committed:
Merge branch 'ornl-next' into 37248-remove-qtassistant-and-reduce-size-of-mantiddocs-package

darshdinger committed:
Merge branch 'ornl-next' into 37248-remove-qtassistant-and-reduce-size-of-mantiddocs-package

darshdinger committed:
Updates to make it all work

darshdinger commented:
Invalid PR, closing
